providers: support "reload"

equalsraf opened the pull request:
### Problem

eval_has_provider() caches its results using static variables for each
provider. Most providers in neovim do not support changes after startup,
however the clipboard provider can be customised later using
g:clipboard. Because eval_has_provider() caches its result,
when no provider is available at startup the clipboard provider will
never work.
This manifests as an inconsistency between :checkhealth that may
indicate a clipboard is available, while getreg('+') still fails.

### Approaches

- ~~add a function to reload the clipboard provider by clearing
the has_clipboard static variable and calling eval_has_provider.
The function is mostly useful for external clipboard providers setup via
g:clipboard that want a delayed setup. So far this is only used for late
UI attachment.~~
- ~~Add a function to skip the static variables check and "reload" the provider.~~
- Remove provider status variables from C and move them to the VimL provider code - with this in place provider reloading should be easier to implement with a reload() function

### Testing

The easiest way to test this is to override the path to prevent nvim from initializing the clipboard provider

```
PATH=/bin bin/nvim --headless -c 'echo $NVIM_LISTEN_ADDRESS'
```

Then attach to nvim using `nvim-qt --server /tmp/nvim/socket` and try

- `:checkhealth`
- `:echo getreg('+')`

nvim-qt provides a function `GuiClipboard()` to setup the clipboard but `getreg()` will still fail unless `reloadprovider()` is called.


### References

- https://github.com/equalsraf/neovim-qt/issues/298
- https://gist.github.com/equalsraf/e8e50a9d503322b7cf52c43c7685e401

bfredl commented:
Should we make this generic over all providers? So `reloadprovider("clipboard")` instead?

equalsraf commented:
My first attempt at this was actually to make it generic for all providers with  an extra argument to eval_has_provider() but i really dont understand the implications of reload other providers because eval_has_provider() actually tries to autoload those scripts.

bfredl commented:
@equalsraf  currently `srcipt_host_eval` and `eval_has_provider()` behaves the same with respect to the different provider scripts (except for some weird method name special case for ruby, but I guess we could eliminate that). Nothing specific need to be implied here, just use the same logic for all providers that you currently use for "clipboard". 

>  If i remember correctly python/ruby/node providers set a guard variable so loading again had no effect.

If currently non-reloadable providers _already_ properly guard against be reloaded, then there is even less reason to replicate exactly the same restriction into `eval.c`. it is rather directly counterproductive: if someone later fixes a script to be reloadable they should be able to enable this just by removing the reload guard, rather than _also_ hacking the c code for every improvement to a specific provider.

equalsraf commented:
Fair point, pushing generic code. Will start looking into tests for this.

justinmk commented:
Let's expose this as `provider#foo#reload()` instead of a new VimL function. Instead of static `has_foo` C variables, the flag could be stored in the provider itself and exposed as `provider#foo#loaded()` or `g:provider_foo_loaded` .

equalsraf commented:
@justinmk  if I'm reading this correctly

eval_has_provider() becomes generic and only calls provider functions, e.g.

```C
bool eval_has_provider(const char *name)
{
  size_t fnamesize = STRLEN(name) + sizeof("provider##enabled") + 1;
  char *fname = xmalloc(fnamesize);
  snprintf(fname, fnamesize, "provider#%s#enabled", name);
  // This is wrong as it prints error messages
  int enabled = call_func_retnr(fname, 0, NULL, true);
  xfree(fname);

  return (enabled > 0);
}
```

Meanwhile all providers define a function provider#foo#enabled() that is always defined on loading and returns true if the provider is available (lets say 1).

provider#foo#reload() is left up to the provider to import or not.

justinmk commented:
@equalsraf yes, does that sound reasonable? This moves more out of the C code and also keeps the interface somewhat consistent (instead of having some `provider#foo` stuff and some special-case VimL functions, we only have `provider#foo` stuff).

equalsraf commented:
Sounds good to me, it also gets rid of inconsistencies with ruby vs others.

My remaining concern here would be performance. eval_has_provider() is called in lots of places

- whenever we call a provider
- for the clipboard is called on middle mouse clicks
- has() is wired directly into eval_has_provider()

No idea how this will behave with autoload. Lets find out.

bfredl commented:
You could turn into a `g:provider#xx#enabled`  variable if you are worried about performance, reading a `g:` var should be a lot cheaper than calling a vimL function repeated times. `provider#xx#reload()` can then change the value, and the name indicates that it shouldn't be changed outside that module (it gives a warning with some verbosity lever).

equalsraf committed:
provider: let providers decide their status

Instead of deciding provider status in eval_has_provider, move the
decision to the provider Vim scripts.

Previously, provider loading worked as follows:

1. eval_has_provider() verified provider availability by searching for
   the provider#providername#Call function and cached this verificaion as a static
   variable for some providers
2. providers short-circuited on loading to prevent the definition of the
   Call function (with the exception of the node provider that did not)

This commit changes the expected interface between nvim and its
providers to facilitate provider reloading, by splitting the
verification of the provider from the availability of the Call function.

eval_has_provider() now checks for a provider#providername#enabled
variable. It is up to the provider script to set this to 0 or 1
accordingly. eval_call_provider() remains unchanged.

All providers hosting a Call function were updated to respect this.

The clipboard provider now has a Reload function to reload the
provider.

equalsraf commented:
> You could turn into a g:provider#xx#enable

This was exactly what i did :D

Also it turns out that all scripts except clipboard do have a load guard that sets some var in g:. So it should be ok.

One thing that is somewhat confusing is how we arrived to the ruby special case in eval_has_provider()
The ruby provider (unlike the others) did no short circuit, instead it always defined Call and provided a Detect function, which sounds a lot like what we want. The node provider is a mix of both, it does not short circuit, and it does have a Detect method (but this was not called from C).

Need to look at the tests later.

justinmk commented:
> One thing that is somewhat confusing is how we arrived to the ruby special case in eval_has_provider() The ruby provider (unlike the others) did no short circuit, instead it always defined Call and provided a Detect function, which sounds a lot like what we want.

AFAIK this is simply a case where no one has had the time to scrutinize whether it makes sense to unify the approaches. I agree the ruby provider approach seems more elegant.

bfredl commented:
`test/functional/fixtures/autoload/provider/clipboard.vim` need to be updated to set enabled

blueyed commented:
Great.  I assume this helps with fixing/adding https://github.com/neovim/neovim/issues/8653 then.

equalsraf commented:
weird ci, did i break the clipboard on mac?

blueyed commented:
For reference:

> From test_quotestar.vim:
Found errors in Test_quotestar():
Run 1:
function RunTheTest[35]..Test_quotestar[6]..Do_test_quotestar_for_macunix line 12: Expected 'text from pasteboard to vim via quotestar' but got ''
function RunTheTest[35]..Test_quotestar[6]..Do_test_quotestar_for_macunix line 19: Expected 'text from vim to pasteboard via quotestar' but got 'text from pasteboard to vim via quotestar'
Run 2:
function RunTheTest[35]..Test_quotestar[6]..Do_test_quotestar_for_macunix line 12: Expected 'text from pasteboard to vim via quotestar' but got ''
function RunTheTest[35]..Test_quotestar[6]..Do_test_quotestar_for_macunix line 19: Expected 'text from vim to pasteboard via quotestar' but got 'text from pasteboard to vim via quotestar'
Flaky test failed too often, giving up
TEST FAILURE

justinmk commented:
LGTM. is this ready?

justinmk commented on runtime/autoload/provider/clipboard.vim:
```diff
@@ -48,6 +48,9 @@ endfunction
 let s:err = ''
 
+" eval_has_provider checks the variable to verify provider status
+let provider#clipboard#enabled = 0
```
```suggestion
let g:provider#clipboard#enabled = 0
```

same for other cases

justinmk commented on runtime/autoload/provider/python3.vim:
```diff
@@ -8,8 +8,10 @@ if exists('g:loaded_python3_provider')
+let provider#python3#enabled = 0
 
 let [s:prog, s:err] = provider#pythonx#Detect(3)
+let provider#python3#enabled = !empty(s:prog)
```
don't really need the declaration on line 11

justinmk commented on src/nvim/eval.c:
```diff
@@ -23985,50 +23985,23 @@ typval_T eval_call_provider(char *provider, char *method, list_T *arguments)
+    return tv.vval.v_number != 0;
+  } else {
+    return true;
+  }
```
ternary

justinmk commented on src/nvim/eval.c:
```diff
@@ -23985,50 +23985,23 @@ typval_T eval_call_provider(char *provider, char *method, list_T *arguments)
+
+  typval_T tv;
+  if (get_var_tv(varname, varname_len, &tv, NULL, false, false) == FAIL) {
+    return false;
```
The old caching mechanism was to avoid searching all of the autoload locations over and over in case the symbol is never defined. 

Now if `provider#%s#enabled` is not defined after searching autoload, we should probably throw an error, except if 'loadplugins' (aka `p_lpl`, aka `--noplugin`) is false. Because all providers should always define this.

equalsraf replied:
has_provider() is called by has() so `has('amiga')` would cause an error. Likewise for all kinds of startup checks.

justinmk replied:
Autoload would be triggered in any case, so we can check that `provider#x#Call` exists first.

justinmk replied:
> The old caching mechanism was to avoid searching all of the autoload locations over and over in case the symbol is never defined.
> 
> Now if `provider#%s#enabled` is not defined after searching autoload, we should probably throw an error, except if 'loadplugins' (aka `p_lpl`, aka `--noplugin`) is false. Because all providers should always define this.

Sorry,  `:help autoload` suggests that I was wrong:

```
This also works when reading a variable that has not been set yet:

        :let l = foo#bar#lvar

However, when the autoload script was already loaded it won't be loaded again
for an unknown variable.
```

justinmk replied:
> has('amiga')

Somewhat related:  it seems that `eval_has_provider()` is expected to trigger autoload. So after this PR, `has('amiga')` will always search for `provider#amiga` on the runtimepath, every time it is called. 

Not sure how to avoid that. Maybe if there's an internal table of known plugin/package names, then anything not in the "provider" key could be skipped by `eval_has_provider()` ...

bfredl replied:
I think we should keep a whitelist with the known provider names. It seems like an anti-feature to me that user code can turn any `has("xyz")` to true. We can easily add more names to the list as we add more provider features.

justinmk replied:
Ok, let's do that for now.

justinmk commented on test/functional/provider/provider_spec.lua:
```diff
@@ -0,0 +1,19 @@
+  it('must set the enabled variable or fail', function()
+    eq(42, eval("provider#brokenenabled#Call('dosomething', [])"))
+	feed_command("call has('brokenenabled')")
+    eq(0, eval("has('brokenenabled')"))
```
is the `feed_command` (line 12) necessary? Has weird indentation also.

justinmk commented:
I will push a proposed change: 

- instead of a new `g:provider#xx#enabled` concept, re-use the `g:loaded_xx_provider` concept. 
    - It's ugly, but is already a very old and common convention, so if we can avoid a new concept then we should.
    - For python providers,  we already have users that look at `g:loaded_python_provider={0,1}` to decide if the python provider was loaded. Adding a new `g:provider#python#enabled` variable to check will just be strange.

- I think `tv_get_number` is needed instead of `get_var_tv`

justinmk committed:
provider: decide status by g:loaded_xx_provider

bfredl commented:
> Instead of a new g:provider#xx#enabled concept, use the existing g:loaded_xx_provider flag. For python providers, we already have users that look at g:loaded_python_provider={0,1} to decide if the python provider was loaded.

But this is not at all what `g:loaded_xx_provider` means and how it has been used so far. Currently it indicates if nvim has already _tried_ to load the provider, and it will set regardless if it worked or not. Users have been using `let g:loaded_python_provider = 1` to stop nvim from even trying. 

So if respecting conventions is important, we should _not_ use it to instead mean whether the provider is successfully started.  Also `g:provider#python#enabled` is not expected to be read/set by any user, they would use `has("python")` as they _always_ have if they care about if the provider is successful. Adding a new variable is certainly less confusing (and doesn't break any conventions) than change the meaning of an existing variable.

> Instead, reload works like this: 

This will also work without changing the meaning of `g:loaded_python_provider`

justinmk commented:
> Currently it indicates if nvim has already tried to load the provider, and it will set regardless if it worked or not. 

Yes and that still works, because the check is only for existence.

> Users have been using let g:loaded_python_provider = 1 to stop nvim from even trying.

True, that's a problem (furthermore, even the help suggests `g:loaded_python_provider=1`, which is wrong). To avoid that problem, the provider could set it to `2` so that `:checkhealth` and `eval_has_provider` know it wasn't set from a vimrc.

>  we should not use it to instead mean whether the provider is successfully started

But it already is used for that, [to some extent](https://github.com/neovim/neovim/issues/6107#issuecomment-279532143). That is why `exists()` is used [instead of checking TRUE](https://github.com/neovim/neovim/commit/9451647d9b9218f679afdf26edddab383b0898a7#diff-8a1575ab1d8124f012f64c63cdaf6de3R7).

bfredl commented:
So the new values are 0/unset (not tried), 1 (user inhibited or failure) and 2 (sucess)? A bit messy, but I think it should work.

justinmk committed:
provider: g:loaded_xx_provider=2 means "enabled and working"

Value of 1 cannot be used, because users might set that in their vimrc
to _disable_ a provider, which would confuse :checkhealth and has().

justinmk commented:
> So the new values are 0/unset (not tried), 1 (user inhibited or failure) and 2 (sucess)? 

Technically zero is the correct user-inhibited way, but we also support 1 for back-compat. See latest commit.

> A bit messy, but I think it should work.

Agreed, but is mostly "internal". It's a tradeoff but I suspect the cognitive burden is less if we avoid an extra variable (especially autoload-style variable, which is not used in any standard plugins).

bfredl commented:
Yes, I was just considering a provider that could enable/disable itself without reload (consider an X11 plugin reacting to changes to `$DISPLAY`), it would be a bit weird to change a "loaded" var without loading anything. But with your last push this is mostly aesthetical, it should not break any user code anymore :)

justinmk committed:
provider: skip non-provider has() feature-names

We don't want to retry autoload sourcing (slow) for every random has()
query that finds it way to eval_call_provider().

justinmk committed:
health.vim: check has("debug")

justinmk committed:
provider: check #Call() if g:loaded_xx_provider=2

justinmk commented:
After CI I think this is ready.

bfredl commented on runtime/doc/develop.txt:
```diff
@@ -101,29 +100,28 @@ implemented in Vim and are now decoupled from Nvim core as providers:
-Python host isn't installed then the plugin will "think" it is running in
-a Vim compiled without the "+python" feature.
+"autoload/provider/python.vim" script, which sets `g:loaded_python_provider`
+to TRUE only if a valid external Python host is found.  Then `has("python")`
```
to 2
